java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -62,89 +64,62 @@ public void scanFile(JavaFileScannerContext context) {
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.TYPE_CAST, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+    return ImmutableList.of(Tree.Kind.TYPE_CAST);
   }
 
   @Override
   public void visitNode(Tree tree) {
-    if (tree.is(Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS)) {
-      addArgsToExclusion(tree);
-    } else {
-      TypeCastTree typeCastTree = (TypeCastTree) tree;
-      JavaType cast = (JavaType) typeCastTree.type().symbolType();
-      ExpressionTree expression = typeCastTree.expression();
-      if (isChainedCastWithWildcard(typeCastTree)) {
-        excluded.add(expression);
-      }
-      if (!excluded.contains(tree)) {
-        JavaType expressionType = (JavaType) expression.symbolType();
-        if (!isExcluded(cast) && (isRedundantNumericalCast(cast, expressionType) || isRedundantCast(cast, expressionType))) {
-          reportIssue(typeCastTree.type(), "Remove this unnecessary cast to \"" + cast + "\".");
-        }
-      }
+    if(!hasSemantic()) {
+      return;
+    }
+    TypeCastTree typeCastTree = (TypeCastTree) tree;
+    Type cast = typeCastTree.type().symbolType();
+    Type target = targetType(typeCastTree);
+    if(target != null && (isRedundantNumericalCast(cast, typeCastTree.expression().symbolType()) || isSubtype(typeCastTree.expression().symbolType(), target))) {
+      reportIssue(typeCastTree.type(), "Remove this unnecessary cast to \"" + cast + "\".");
     }
   }
 
-  private static boolean isChainedCastWithWildcard(TypeCastTree typeCastTree) {
-    ExpressionTree expression = typeCastTree.expression();
-    return expression.is(Tree.Kind.TYPE_CAST) && usesWildCard(expression);
-  }
-
-  private void addArgsToExclusion(Tree tree) {
-    List<ExpressionTree> args;
-    if(tree.is(Tree.Kind.METHOD_INVOCATION)) {
-      MethodInvocationTree mit = (MethodInvocationTree) tree;
-      args = mit.arguments();
-    } else {
-      NewClassTree newClassTree = (NewClassTree) tree;
-      args = newClassTree.arguments();
-    }
-    for (ExpressionTree arg : args) {
-      if (arg.is(Tree.Kind.TYPE_CAST)) {
-        excluded.add(arg);
+  @CheckForNull
+  private Type targetType(TypeCastTree tree) {
+    Tree parent = skipParentheses(tree.parent());
+    Type target = null;
+    if(parent.is(Tree.Kind.RETURN_STATEMENT)) {
+      Tree method = parent;
+      while (!method.is(Tree.Kind.METHOD)) {
```

Comment:
Return statement aren't necessary in a method. The following case will trigger a NPE.
```
class H {
  Callable<H> c0 = () -> {
    return (H) getValue();
  };
  
  private Object getValue() {
    return null;
  }
}
```